README
Inotify is on erlang port for the Linux inotify API allowing one to monitor
changes to files and directory in the filesystem.

Documentation
-------------

Most documentation is here:
http://sheyll.github.com/inotify/

For an example refer to the unit test:
https://github.com/sheyll/inotify/blob/master/test/inotify_test.erl


You might want to have the documentation locally, this can be achieved by

$ ./rebar doc

which generates HTML documentation in the "doc" sub directory.


License
-------
In short, you can do anything you want with the code including using it as part
of you plan for world domination (if your successful can I have one of the nicer
countries please). No responsiblity it taken for the fitness of the any purpose,
etc, etc. The only thing I ask is that if you find a bug and fix send me the
patch. Likewise, feature suggestions and patches are welcome.

TODO
----
* add support for multiple controller functions
* Write some documentation!

Release History
---------------
20130101 version 0.4.0 switch to rebar
20100206 version 0.3 on github
20090221 release 0.2 bug fix
20080929 initial release version 0.1


ChangeLog
2013-01-01  Sven Heyll <sven.heyll@gmail.com>

        * Move src/test.erl to priv/test.erl

        * Remove auto tools support

        * Add rebar configuration

        * Add OTP Application

        * Add OTP gen_event     

2010-02-01  Mats Cronqvist  <masse@kreditor.se>

	* c_src/inotify_erlang.c (note_remove):
	driver will always return either a 2-tuple (synchronously) or a
	5-tuple (asynchronously)

	* src/inotify.erl (test_start):
	see above

	* src/inoteefy.erl (talk_to_port):
	fixed race condition between synchronous and asynchronous events
	



test/inotify_test.erl
%%%=============================================================================
%%% @author Sven Heyll <sven.heyll@lindenbaum.eu>
%%% @copyright (C) 2013, Sven Heyll
%%%=============================================================================

-module(inotify_test).

-export([inotify_event/3]).

-include_lib("eunit/include/eunit.hrl").
-include("inotify.hrl").

%%%=============================================================================
%%% TESTS
%%%=============================================================================

monitor_file_test() ->
    process_flag(trap_exit, true),
    file:delete("/tmp/inotify_test"),

    inotify:start(x,y),
    Ref1 = inotify:watch("/tmp/", ?ALL),
    ok = inotify:add_handler(Ref1, ?MODULE, self()),
    receive after 100 -> ok end,

    {ok, H} = file:open("/tmp/inotify_test", [read, write]),

    receive
        {[create],0,"inotify_test"} -> pass;
        M1 -> throw({expected_message_file_create_but_got, M1})
    after 1000 -> throw(expected_create)
    end,
    receive
        {[open],0,"inotify_test"} -> pass;
        M2 -> throw({expected_message_file_open_but_got, M2})
    after 1000 -> throw(expected_open)
    end,
    Ref2 = inotify:watch("/tmp/inotify_test"),
    inotify:unwatch(Ref1),
    ok = inotify:add_handler(Ref2, ?MODULE, self()),
    receive after 100 -> ok end,
    io:write(H, "test"),
    receive
        {[modify],0,""} -> pass;
        M3 -> throw({expected_message_file_modify_but_got, M3})
    after 1000 -> throw(expected_modify)
    end,
    ok = file:close(H),
    receive
        {[close_write],0,""} -> pass;
        M4 -> throw({expected_message_file_close_write_but_got, M4})
    after 1000 -> throw(expected_close_write)
    end,
    Ref3 = inotify:watch("/tmp/", ?ALL),
    ok = inotify:add_handler(Ref3, ?MODULE, self()),
    receive after 100 -> ok end,
    ok = file:delete("/tmp/inotify_test"),
    receive
        {[attrib],0,""} -> pass
    after 1000 ->
            receive X1 -> throw({expected, X1})
            after 1000 -> throw(expected_attrib)
            end
    end,
    receive
        {[delete_self],0,""} -> pass
    after 1000 ->
            receive X2 -> throw({expected, X2})
            after 1000 -> throw(expected_delete_self)
            end
    end,
    receive
        {[ignored],0,""} -> pass
    after 1000 ->
            receive X3 -> throw({expected, X3})
            after 1000 -> throw(expected_ignored)
            end
    end,
    receive
        {[delete],0,"inotify_test"} -> pass
    after 1000 ->
            receive X4 -> throw({expected, X4})
            after 1000 -> throw(expected_delete)
            end
    end.

inotify_event(TestPid, Ref, Msg = ?inotify_msg(M, C, N)) ->
    io:format("$$$ ~p: ~p~n", [Ref, Msg]),
    TestPid ! {M, C, N},
    ok.


.eunit/inotify_test.erl
%%%=============================================================================
%%% @author Sven Heyll <sven.heyll@lindenbaum.eu>
%%% @copyright (C) 2013, Sven Heyll
%%%=============================================================================

-module(inotify_test).

-export([inotify_event/3]).

-include_lib("eunit/include/eunit.hrl").
-include("inotify.hrl").

%%%=============================================================================
%%% TESTS
%%%=============================================================================

monitor_file_test() ->
    process_flag(trap_exit, true),
    file:delete("/tmp/inotify_test"),

    inotify:start(x,y),
    ok = inotify:watch("/tmp/", tag1, ?ALL),
    ok = inotify:add_handler(tag1, ?MODULE, self()),
    receive after 100 -> ok end,

    {ok, H} = file:open("/tmp/inotify_test", [read, write]),

    receive
        {[create],0,"inotify_test"} -> pass;
        M1 -> throw({expected_message_file_create_but_got, M1})
    after 1000 -> throw(expected_create)
    end,
    receive
        {[open],0,"inotify_test"} -> pass;
        M2 -> throw({expected_message_file_open_but_got, M2})
    after 1000 -> throw(expected_open)
    end,
    inotify:watch("/tmp/inotify_test", tag2),
    inotify:unwatch(tag1),
    ok = inotify:add_handler(tag2, ?MODULE, self()),
    receive after 100 -> ok end,
    io:write(H, "test"),
    receive
        {[modify],0,""} -> pass;
        M3 -> throw({expected_message_file_modify_but_got, M3})
    after 1000 -> throw(expected_modify)
    end,
    ok = file:close(H),
    receive
        {[close_write],0,""} -> pass;
        M4 -> throw({expected_message_file_close_write_but_got, M4})
    after 1000 -> throw(expected_close_write)
    end,
    ok = inotify:watch("/tmp/", tag3, ?ALL),
    ok = inotify:add_handler(tag3, ?MODULE, self()),
    receive after 100 -> ok end,
    ok = file:delete("/tmp/inotify_test"),
    receive
        {[attrib],0,""} -> pass
    after 1000 ->
            receive X1 -> throw({expected, X1})
            after 1000 -> throw(expected_attrib)
            end
    end,
    receive
        {[delete_self],0,""} -> pass
    after 1000 ->
            receive X2 -> throw({expected, X2})
            after 1000 -> throw(expected_delete_self)
            end
    end,
    receive
        {[ignored],0,""} -> pass
    after 1000 ->
            receive X3 -> throw({expected, X3})
            after 1000 -> throw(expected_ignored)
            end
    end,
    receive
        {[delete],0,"inotify_test"} -> pass
    after 1000 ->
            receive X4 -> throw({expected, X4})
            after 1000 -> throw(expected_delete)
            end
    end.

inotify_event(TestPid, Tag, Msg = ?inotify_msg(M, C, N)) ->
    io:format("$$$ ~p: ~p~n", [Tag, Msg]),
    TestPid ! {M, C, N},
    ok.


src/inotify.erl
%%%-------------------------------------------------------------------
%%% Created : 28 Jan 2010 by Mats Cronqvist <masse@kreditor.se>
%%% Converted to an OTP application by Sven Heyll <sven.heyll@gmail.com>
%%%
%%% @doc
%%% Application that monitors files via Linux `inotify'. A user can tell
%%% `inotify' to {@link watch/2} for specific events on several
%%% files/directories.  To actually get any events the User must call {@link
%%% add_handler/3}.
%%%
%%% To {@link unwatch/1} something the User must pass the return value of {@link
%%% watch/2} which is a unique reference to the monitor create by {@link
%%% watch/2}.
%%%
%%% This module also defines a behaviour, with a single function
%%% `inotify_event(Arg, EventRef, Msg)'
%%%
%%% A callback for file monitoring event handler added by {@link add_handler/3}.
%%%
%%% `Arg' is the user provided extra argument.
%%%
%%% `EventRef' is the value returned by {@link watch/2}.
%%%
%%% The `Msg' parameter of the callback function should be pattern-matched
%%% with the macro `?inotify_msg(Mask, Cookie, OptionalName)' contained
%%% in `include/inotify.hrl'.
%%%
%%% `Mask' is a list of atoms describing the nature of the event. Refer to the
%%% Linux inotify man page for a detailed description.
%%%
%%% `Cookie' is 0 except when a file is moved, where it is used to identify
%%% `move_from' and `move_to' events belonging to the same move-operation.
%%%
%%% `OptionalName' contains the relative file name when monitoring whole
%%% directories. E.g. when monitoring "/tmp" the following event might be
%%% created when opening "/tmp/xxx": `?inotify_msg([open], 0, "xxx")'.
%%% @end
%%%-------------------------------------------------------------------

-module(inotify).

-behaviour(application).
-behaviour(supervisor).

%% API
-export([watch/1, watch/2, unwatch/1, add_handler/3, print_events/1]).

%% Application callbacks
-export([start/2, stop/1]).

%% Supervisor callbacks
-export([init/1]).

%% inotify_evt callbacks
-export([inotify_event/3]).

-export_type([mask/0, msg/0]).

-define(SERVER, inotify_server).
-define(EVENT, inotify_evt).

-include("inotify.hrl").

-type msg() :: ?inotify_msg(Mask :: [mask()],
                            Cookie :: non_neg_integer(),
                            OptionalName :: string()).
%% A file monitoring message.

-type mask() :: ?ALL           |
                ?ACCESS        |
                ?ATTRIB        |
                ?CLOSE_WRITE   |
                ?CLOSE_NOWRITE |
                ?CLOSE         |
                ?CREATE        |
                ?DELETE        |
                ?DELETE_SELF   |
                ?MODIFY        |
                ?MOVE_SELF     |
                ?MOVED_FROM    |
                ?MOVED_TO      |
                ?MOVE          |
                ?OPEN          |
                ?DONT_FOLLOW   |
                ?MASK_ADD      |
                ?ONLYDIR.
%% The type of a file monitoring event.

%%%===================================================================
%%% Behaviour definition
%%%===================================================================

%%--------------------------------------------------------------------
%% FUCK YOU EDOC!
%%--------------------------------------------------------------------
-callback inotify_event(Arg :: term(),
                        EventRef :: term(),
                        Msg :: inotify:msg()) ->
    ok | remove_handler.

%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc
%% Add a file/directory monitor for all file events. Events for the
%% file/directory will be published via a {@link inotify_evt}. The events
%% resulting from this call will be identified by the unique reference returned
%% by this function, no matter if the operation was successful or not.
%%
%% This reference must also be passed to {@link unwatch/1}.
%% @see inotify:add_handler/3
%% @see inotify:unwatch/1
%% @end
%%--------------------------------------------------------------------
-spec watch(string()) ->
                   reference().
watch(File) ->
    watch(File, ?ALL).

%%--------------------------------------------------------------------
%% @doc
%% Add a file/directory monitor for specific events.  This differs from {@link
%% watch/2} by the additional `mask' argument.  Calling {@link watch/2} is
%% equivalant to calling {@link watch/3} with `?ALL' as second parameter.
%%
%% @see inotify:watch/1
%% @end
%%--------------------------------------------------------------------
-spec watch(string(), mask() | [mask()]) ->
                   reference().
watch(File, Mask) ->
    EventTag = make_ref(),
    inotify_server:watch(File, EventTag, Mask),
    EventTag.

%%--------------------------------------------------------------------
%% @doc
%% Remove a monitor added via {@link watch/2}, `Ref' identifies the
%% monitor.
%% @end
%%--------------------------------------------------------------------
-spec unwatch(reference()) ->
                     ok.
unwatch(Ref) ->
    inotify_evt:remove_all_handlers(Ref),
    inotify_server:unwatch(Ref).

%%--------------------------------------------------------------------
%% @doc
%% Add an event handler that receives all events generated by a file monitor
%% identified by a tag that was passed to {@link inotify:watch/2}. When an event
%% occurs for `Ref' the handler will call `Module:inotify_event(Arg,
%% EventTag, Msg)' as defined by the `callback' definition in {@link
%% inotify_event/3}.
%% @end
%%--------------------------------------------------------------------
-spec add_handler(reference(), module(), term()) ->
                         ok.
add_handler(Ref, Module, Arg) ->
    inotify_evt:add_handler(Ref, Module, Arg).

%%--------------------------------------------------------------------
%% @doc
%% Add an event handler that prints to stdout events. The events to print are
%% identified by a reference returned by {@link watch/2}.
%% @end
%%--------------------------------------------------------------------
-spec print_events(reference()) ->
                          ok.
print_events(Ref) ->
    inotify_evt:add_handler(Ref, ?MODULE, []).


%%%===================================================================
%%% Application callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
start(_StartType, _StartArg) ->
    supervisor:start_link(?MODULE, []).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
stop(_) ->
    ok.

%%%===================================================================
%%% supervisor callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
init([]) ->
  {ok, {{one_for_all, 1, 1},
        [{?EVENT,  {?EVENT, start_link, []},
          permanent, 2000, worker, [?EVENT]},
         {?SERVER, {?SERVER, start_link, []},
          permanent, 2000, worker, [?SERVER]}]}}.

%%%===================================================================
%%% inotify_evt callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc
%% This module actually implements the {@link inotify_evt} behaviour, with a
%% function that simply prints the events to stdout.
%%
%% This function can be used as an example for how to implement a callback.
%%
%% @see print_events/1
%% @end
%%--------------------------------------------------------------------
inotify_event([], Ref, ?inotify_msg(Masks, Cookie, OptionalName)) ->
    io:format("[INOTIFY] - ~p - ~p ~p ~p~n", [Ref, Masks, Cookie, OptionalName]).


src/inotify_evt.erl
%%%-------------------------------------------------------------------
%%% @author Sven Heyll <sven@sheyllpc>
%%% @copyright (C) 2013, Sven Heyll
%%% @doc
%%% Event manager for file monitoring events.
%%% This is an internal module.
%%% @end
%%% Created :  1 Jan 2013 by Sven Heyll <sven@sheyllpc>
%%%-------------------------------------------------------------------
-module(inotify_evt).

-behaviour(gen_event).

%% API
-export([start_link/0, publish/2, add_handler/3, remove_all_handlers/1]).

%% gen_event callbacks
-export([init/1, handle_event/2, handle_call/2,
         handle_info/2, terminate/2, code_change/3]).

-define(SERVER, ?MODULE).

-include("inotify.hrl").

%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
start_link() ->
    gen_event:start_link({local, ?SERVER}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
publish(EventTag, Msg) ->
    gen_event:notify(?MODULE, {?MODULE, EventTag, Msg}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
add_handler(EventTag, Module, Arg) ->
    gen_event:add_handler(?MODULE,
                          {?MODULE, Arg},
                          {EventTag, Module, Arg}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
remove_all_handlers(EventTag) ->
    gen_event:sync_notify(?MODULE, {remove_all_handlers, EventTag}).

%%%===================================================================
%%% gen_event callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
init({EventTag, Module, Arg}) ->
    {ok, {EventTag, Module, Arg}}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_event({remove_all_handlers, EventTag},
             {EventTag, _Module, _Arg}) ->
    remove_handler;

handle_event({?MODULE, EventTag, Msg = ?inotify_msg(_, _, _)},
             State = {EventTag, Module, Arg}) ->
    case Module:inotify_event(Arg, EventTag, Msg) of
        ok ->
            {ok, State};
        remove_handler ->
            remove_handler
    end;

handle_event(_Event, State) ->
    {ok, State}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_call(_Request, State) ->
    Reply = ok,
    {ok, Reply, State}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_info(_Info, State) ->
    {ok, State}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
terminate(_Reason, _State) ->
    ok.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%%===================================================================
%%% Internal functions
%%%===================================================================


src/inotify_server.erl
%%%-------------------------------------------------------------------
%%% Created : 28 Jan 2010 by Mats Cronqvist <masse@kreditor.se>
%%% Converted to an OTP application by Sven Heyll <sven.heyll@gmail.com>
%%%
%%% @doc
%%% A `gen_server' that provides access to low level Linux `inotify'.
%%% This is an internal module. Use {@link inotify} instead.
%%% @end
%%%-------------------------------------------------------------------

-module(inotify_server).

-behaviour(gen_server).

-author('Mats Cronqvist').
-author('Sven Heyll').

%% API
-export([start_link/0,
         watch/2,
         watch/3,
         unwatch/1]).

%% gen_server exports
-export([init/1,
         terminate/2,
         handle_cast/2,
         handle_info/2]).

-define(SERVER, ?MODULE).

-include("inotify.hrl").

-record(ld,{port, fd}).

-define(log(T),
        error_logger:info_report(
          [process_info(self(), current_function),
           {line, ?LINE},
           T])).

%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
start_link() ->
  gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
watch(File, EventTag) ->
    gen_server:cast(?SERVER, {watch, {File, EventTag, ?ALL}}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
watch(File, EventTag, Mask) ->
    gen_server:cast(?SERVER, {watch, {File, EventTag, Mask}}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
unwatch(EventTag) ->
    gen_server:cast(?SERVER, {unwatch, EventTag}).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
init([]) ->
  process_flag(trap_exit, true),
  Port = open_port(),
  {ok, FD} = talk_to_port(Port , {open}),
  {ok, #ld{port = Port, fd = FD}}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
terminate(_, LD) ->
  talk_to_port(LD#ld.port, {close, LD#ld.fd}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_info({Port, {data,Msg}}, LD = #ld{port = Port}) ->
  maybe_call_back(binary_to_term(Msg)),
  {noreply, LD};

handle_info(Msg, LD) ->
  ?log({unknown_message, Msg}),
  {noreply, LD}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_cast(stop, LD) ->
  {stop,normal, LD};

handle_cast({watch,Watch},LD) ->
  {noreply, do_watch(Watch, LD)};

handle_cast({unwatch, Unwatch}, LD) ->
  {noreply, do_unwatch(Unwatch, LD)};

handle_cast(Msg, LD) ->
  ?log({unknown_message, Msg}),
  {noreply, LD}.

%%%===================================================================
%%% Internal functions
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
open_port() ->
    E = filename:join([code:priv_dir(inotify), "inotify"]),
    open_port({spawn, E},[{packet, 2}, binary, exit_status]).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
maybe_call_back({event, WD, Mask, Cookie, Name}) ->
    case get({wd, WD}) of
        undefined ->
            case Mask of
                [ignored] ->
                    ok;
                _ ->
                    ?log([{got_event_without_callback, WD, Mask, Cookie, Name}])
            end;
        Tag ->
            inotify_evt:publish(Tag, ?inotify_msg(Mask, Cookie, Name))
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
do_watch({File, Tag, Mask},LD) ->
    try
        {ok, WD} = talk_to_port(LD#ld.port, {add, LD#ld.fd, File, Mask}),
        put({tag, Tag}, WD),
        put({wd, WD}, Tag),
        LD
    catch
        C:R ->
            ?log([{error_watching_file, File, Tag}, {C, R}]),
            LD
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
do_unwatch(Tag, LD) ->
    case get({tag, Tag}) of
        undefined ->
            ?log([{not_watching, Tag}]),
            LD;
        WD ->
            try
                talk_to_port(LD#ld.port, {remove, LD#ld.fd, WD})
            catch
                C:R ->
                    ?log([{error_unwatching, Tag},{C, R}])
            end,
            erase({tag, Tag}),
            erase({wd, WD}),
            LD
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
talk_to_port(Port,Msg) ->
    try
        erlang:port_command(Port, term_to_binary(Msg)),
        receive
            {Port, {data, D = <<131, 104, 2, _/binary>>}} ->
                binary_to_term(D)
        after 1000 ->
                throw(talk_to_port_timeout)
        end
    catch
        _:R ->
            throw({talking_to_port_failed, {R, Port, Msg}})
    end.
